File: wpf/src/Image/Thumbnails.cs
/*
MIT License

Copyright (c) 2021 Max Kas

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

using Mku.Salmon.Streams;
using Salmon.Vault.ViewModel;
using Salmon.Vault.Utils;
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.IO;
using System.Reflection;
using System.Security.Cryptography;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Media;
using System.Windows.Media.Imaging;
using BitConverter = Mku.Convert.BitConverter;
using Mku.SalmonFS.File;
using Mku.FS.Drive.Utils;
using Mku.FS.File;
using System.Diagnostics;
using System.Threading;

namespace Salmon.Vault.Image;

/// <summary>
/// Utility for thumbnail generation
/// </summary>
public class Thumbnails
{
    private static readonly int TMP_VIDEO_THUMB_MAX_SIZE = 5 * 1024 * 1024;
    private static readonly int TMP_GIF_THUMB_MAX_SIZE = 512 * 1024;
    private static readonly int BUFFER_SIZE = 256 * 1024;
    private static readonly int THUMBNAIL_SIZE = 128;
    private static readonly int MAX_CACHE_SIZE = 20 * 1024 * 1024;

    private static readonly ConcurrentDictionary<AesFile, BitmapImage> cache = new ConcurrentDictionary<AesFile, BitmapImage>();
    private static int cacheSize;
    private static object _lock = new object();

    /// <summary>
    /// Returns a bitmap thumbnail from an encrypted file
    /// </summary>
    /// <param name="salmonFile">The encrypted media file which will be used to get the thumbnail</param>
    /// <returns></returns>
    public static BitmapImage GetVideoThumbnail(AesFile salmonFile)
    {
        return GetVideoThumbnail(salmonFile, 0);
    }

    public static BitmapImage GetVideoThumbnail(AesFile salmonFile, long ms)
    {
        throw new NotSupportedException();
    }

    public static BitmapImage GetVideoThumbnailMedia(IFile file, long ms)
    {
        throw new NotSupportedException();
    }

    /// <summary>
    /// Create a partial temp file from an encrypted file that will be used to get the thumbnail
    /// </summary>
    /// <param name="salmonFile">The encrypted file that will be used to get the temp file</param>
    /// <returns></returns>
    private static IFile GetVideoTmpFile(AesFile salmonFile)
    {
        throw new NotSupportedException();
    }

    /// <summary>
    /// Return a MemoryStream with the partial unencrypted file contents.
    /// This will read only the beginning contents of the file since we don't need the whole file.
    /// </summary>
    /// <param name="salmonFile">The encrypted file to be used</param>
    /// <param name="maxSize">The max content length that will be decrypted from the beginning of the file</param>
    /// <returns></returns>
    private static Stream GetTempStream(AesFile salmonFile, long maxSize)
    {
        MemoryStream ms = new MemoryStream();
        AesStream ins = salmonFile.GetInputStream();
        byte[] buffer = new byte[BUFFER_SIZE];
        int bytesRead;
        long totalBytesRead = 0;
        while ((bytesRead = ins.Read(buffer, 0, buffer.Length)) > 0
                && totalBytesRead < maxSize)
        {
            ms.Write(buffer, 0, bytesRead);
            totalBytesRead += bytesRead;
        }
        ms.Flush();
        ins.Close();
        ms.Position = 0;
        return ms;
    }

    /// <summary>
    /// Create a bitmap from the unencrypted data contents of a media file
    /// If the file is a gif we get only a certain amount of data from the beginning of the file
    /// since we don't need to get the whole file.
    /// </summary>
    /// <param name="salmonFile"></param>
    /// <returns></returns>
    public static void GenerateThumbnailAsync(SalmonFileViewModel item)
    {
        if (cache.ContainsKey(item.GetAesFile()))
        {
            BitmapImage bitmapImage = cache[item.GetAesFile()];
            WindowUtils.RunOnMainThread(() =>
            {
                item.Image = bitmapImage;
            });
            return;
        }
        Task.Run(() =>
        {
            
            try
            {
                lock (_lock)
                {
                    GenerateThumbnail(item);
                }
            }
            catch (Exception ex)
            {
                Debug.WriteLine(ex);
                Console.Error.WriteLine(ex);
                WindowUtils.RunOnMainThread(() =>
                {
                    SetDefaultIcon(item);
                });
            }
        }).ConfigureAwait(false);
    }

    private static void SetDefaultIcon(SalmonFileViewModel item)
    {
        BitmapImage bitmapImage = GetIcon(item.GetAesFile());
        AddCache(item.GetAesFile(), bitmapImage);
        item.Image = bitmapImage;
    }

    private static void GenerateThumbnail(SalmonFileViewModel item)
    {

        bool isDirectory = false;
        string name = "";
        isDirectory = item.GetAesFile().IsDirectory;
        name = item.GetAesFile().Name;
        
        BitmapImage bitmapImage;
        if (isDirectory || !FileUtils.IsImage(name))
        {
            WindowUtils.RunOnMainThread(() =>
            {
                SetDefaultIcon(item);
            });
        }
        else
        {
            // we might have multiple requests so we make sure we process only once
            AddCache(item.GetAesFile(), null);

            Stream nStream = GenerateThumbnail(item.GetAesFile());
            if(nStream == null)
            {
                SetDefaultIcon(item);
                return;
            }
            WindowUtils.RunOnMainThread(() =>
            {
                try
                {
                    bitmapImage = new BitmapImage();
                    bitmapImage.BeginInit();
                    bitmapImage.CacheOption = BitmapCacheOption.OnLoad;
                    bitmapImage.StreamSource = nStream;
                    bitmapImage.EndInit();
                    bitmapImage.Freeze();
                    item.Image = bitmapImage;
                    AddCache(item.GetAesFile(), bitmapImage);
                }
                catch (Exception ex)
                {
                    Debug.WriteLine(ex);
                    Console.Error.WriteLine(ex);
                    SetDefaultIcon(item);
                }
            });
        }
    }

    private static BitmapImage GetIcon(AesFile salmonFile)
    {
        string icon = salmonFile.IsFile ? "Icons/file_item_small.png" : "Icons/folder_small.png";
        Uri uri = new Uri("pack://application:,,,/"
            + Assembly.GetAssembly(typeof(Thumbnails)).GetName().Name
            + ";component/" + icon, UriKind.RelativeOrAbsolute);
        BitmapImage bitmapImage = new BitmapImage(uri);
        return bitmapImage;
    }

    private static Stream GenerateThumbnail(AesFile file)
    {
        Stream stream = FromFile(file);
        if (stream == null)
            return null;
        System.Drawing.Bitmap bitmap = new System.Drawing.Bitmap(stream);
        stream.Close();
        bitmap = ResizeBitmap(bitmap, THUMBNAIL_SIZE);
        return FromBitmap(bitmap);
    }

    private static Stream FromBitmap(System.Drawing.Bitmap bitmap)
    {
        MemoryStream ms = new MemoryStream();
        bitmap.Save(ms, System.Drawing.Imaging.ImageFormat.Bmp);
        ms.Position = 0;
        return ms;
    }

    private static System.Drawing.Bitmap ResizeBitmap(System.Drawing.Bitmap bitmap, int size)
    {
        float ratio = bitmap.Width / (float)bitmap.Height;
        int nWidth = ratio > 1 ? size : (int)(size * ratio);
        int nHeight = ratio < 1 ? size : (int)(size / ratio);
        return new System.Drawing.Bitmap(bitmap, new System.Drawing.Size(nWidth, nHeight));
    }

    private static void AddCache(AesFile file, BitmapImage image)
    {
        if (cacheSize > MAX_CACHE_SIZE)
            ResetCache();
        cache[file] = image;
        try
        {
            if (image != null)
                cacheSize += (int)(image.Width * image.Height * 4);
        }
        catch (Exception ex)
        {
            Console.WriteLine(ex);
        }
    }

    public static void ResetCache()
    {
        int reduceSize = 0;
        List<AesFile> keysToRemove = new List<AesFile>();
        foreach (AesFile key in cache.Keys)
        {
            BitmapImage bitmap = cache[key];
            if (bitmap != null)
                reduceSize += (int)(bitmap.Width * bitmap.Height * 4);
            if (reduceSize >= MAX_CACHE_SIZE / 2)
                break;
            keysToRemove.Add(key);
        }
        foreach (AesFile key in keysToRemove)
        {
            cache.Remove(key, out BitmapImage bitmap);
            if (bitmap != null)
                cacheSize -= (int)(bitmap.Width * bitmap.Height * 4);
        }
    }

    public static void ResetCache(AesFile file)
    {
        if (cache.ContainsKey(file))
        {
            BitmapImage image = cache[file];
            cacheSize -= (int)(image.Width * image.Height * 4);
            cache.Remove(file, out _);
        }
    }

    private static Stream FromFile(AesFile file)
    {
        Stream stream = null;
        try
        {
            string ext = FileUtils.GetExtensionFromFileName(file.Name).ToLower();
            if (ext.Equals("gif") && file.RealFile.Length > TMP_GIF_THUMB_MAX_SIZE)
            {
                stream = GetTempStream(file, TMP_GIF_THUMB_MAX_SIZE);
            }
            else
            {
                stream = GetTempStream(file, long.MaxValue);
            }
            stream.Position = 0;
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine(ex);
            if (stream != null)
                stream.Close();
            throw ex;
        }
        return stream;
    }

    public static Color GetTintColor(AesFile salmonFile)
    {
        if (!salmonFile.IsFile || FileUtils.IsImage(salmonFile.Name))
            return Colors.Transparent;

        SHA256 sha256 = SHA256.Create();
        string ext = GetExt(salmonFile);
        byte[] hashValue = sha256.ComputeHash(Encoding.UTF8.GetBytes(ext));
        string hashstring = BitConverter.ToHex(hashValue);
        Color color = (Color)ColorConverter.ConvertFromString("#" + hashstring.Substring(0, 6));
        return color;
    }

    public static string GetExt(AesFile salmonFile)
    {
        if (!salmonFile.IsFile || FileUtils.IsImage(salmonFile.Name))
            return "";
        return FileUtils.GetExtensionFromFileName(salmonFile.Name).ToLower();
    }
}

File: js/src/assets/js/vault/image/thumbnails.js
/*
MIT License

Copyright (c) 2021 Max Kas

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

import { FileUtils } from "../../lib/simple-fs/fs/drive/utils/file_utils.js";
import { MemoryStream } from "../../lib/simple-io/streams/memory_stream.js";

/**
 * Utility class that generates thumbnails for encrypted salmon files
 */
export class Thumbnails {
    static TMP_THUMB_DIR = "tmp";
    static TMP_VIDEO_THUMB_MAX_SIZE = 3 * 1024 * 1024;
    static TMP_GIF_THUMB_MAX_SIZE = 1 * 1024 * 1024;
    static ENC_BUFFER_SIZE = 512 * 1024;
    static THUMBNAIL_SIZE = 64;

    static MAX_CACHE_SIZE = 128 * 1024;
    static cache = new Map();
    static TINT_COLOR_ALPHA = 127;
    static cacheSize = 0;
    static enableCache = true;
    static objectURLs = new Set();


    static isAnimationEnabled() {
        return !Thumbnails.animationStopped;
    }

    static animationStopped = false;

    static enableAnimation(value) {
        Thumbnails.animationStopped = !value;
    }

    /**
     * Returns a bitmap thumbnail from an encrypted file
     * @param {AesFile} salmonFile The file
     * @param {number} position The position in seconds
     * @returns {Promise<>}
     */
    static async getVideoThumbnail(salmonFile, position = 3) {
        let blob = await Thumbnails.#getVideoTmpBlob(salmonFile);
        let imageUrl = Thumbnails.createObjectURL(blob);
        return new Promise((resolve, reject) => {
            let video = document.createElement('video');
            video.setAttribute('src', imageUrl);
            video.addEventListener('loadedmetadata', () => {
                video.addEventListener('seeked', () => {
                    const canvas = document.createElement("canvas");
                    canvas.width = video.videoWidth;
                    canvas.height = video.videoHeight;
                    const ctx = canvas.getContext("2d");
                    ctx.drawImage(video, 0, 0, canvas.width, canvas.height);
                    ctx.canvas.toBlob((blob) => {
                        const thumbnailImage = new Image();
                        let thumbnailImageUrl = Thumbnails.createObjectURL(blob);
                        thumbnailImage.src = thumbnailImageUrl;
                        resolve(thumbnailImage);
                    }, 'image/png', 0.75);
                });
                if (video.duration < position)
                    position = 0;
                video.currentTime = position;
            });
            video.load();
        });
    }

    static async #getVideoTmpBlob(salmonFile) {
        let ms = await Thumbnails.#getTempStream(salmonFile, Thumbnails.TMP_VIDEO_THUMB_MAX_SIZE);
        let blob = new Blob([ms.toArray().buffer]);
        await ms.close();
        return blob;    
    }

    static async #getTempStream(salmonFile, maxSize) {
        let ms = new MemoryStream();
        let ins = await salmonFile.getInputStream();
        let buffer = new Uint8Array(Thumbnails.ENC_BUFFER_SIZE);
        let bytesRead;
        let totalBytesRead = 0;
        while ((bytesRead = await ins.read(buffer, 0, buffer.length)) > 0
                && totalBytesRead < maxSize) {
            await ms.write(buffer, 0, bytesRead);
            totalBytesRead += bytesRead;
        }
        await ms.flush();
        await ins.close();
        await ms.setPosition(0);
        return ms;
    }

    /**
     * 
     * @param {AesFile} salmonFile The aes file
     * @param {number} width The width
     * @param {number} height The height
     * @param {number} position The position in seconds if file is media
     * @returns 
     */
    static async generateThumbnail(salmonFile, width, height, position = 3) {
        let image = null;
        let key = await Thumbnails.getHash(salmonFile) + ":" + width + ":" + height;
        if (await salmonFile.isFile() && FileUtils.isVideo(await salmonFile.getName())) {        
            key += ":" + position;
        }
        if (Thumbnails.cache.has(key)) {
            let size = 0;
            [image,size] = Thumbnails.cache.get(key);
            if(image.parentElement!=null)
                image.parentElement.removeChild(image);
            return image;
        }
        
        try {
            if (await salmonFile.isFile() && FileUtils.isImage(await salmonFile.getName())) {
                image = await Thumbnails.getImageThumbnail(salmonFile);
                image = await Thumbnails.#resize(image, width, height);
            } else if (await salmonFile.isFile() && FileUtils.isVideo(await salmonFile.getName())) {
                image = await Thumbnails.getVideoThumbnail(salmonFile, position);
                image = await Thumbnails.#resize(image, width, height);
            }
        } catch (e) {
            throw e;
        }
        if(image != null)
            Thumbnails.addCache(key, image);
        return image;
    }

    static async getIcon(salmonFile, width, height) {
        let icon = await salmonFile.isFile() ? 
        "assets/images/common-res/icons/file_item_small.png" : 
        "assets/images/common-res/icons/folder_small.png";
        let image = new Image();
        if (image.width > image.height) {
            image.width = width;
        } else {
            image.height = height;
        }
        image.src = icon;
        if (await salmonFile.isFile()) {
            try {
                let ext = FileUtils.getExtensionFromFileName(await salmonFile.getName()).toLowerCase();
                let hsv = await Thumbnails.getHSVFromExtension(ext);
                Thumbnails.addImage(image, hsv);
            } catch (ex) {
                console.error(ex);
            }
        }

        return image;
    }

    static addImage(image, hsv) {
        let [h, s, v] = hsv;
        let filter = `sepia(100%) saturate(${Math.trunc((s + 1) * 100)}%) 
            brightness(${Math.trunc((v + 0.4) * 100)}%) hue-rotate(${Math.trunc(h)}deg)`;
        image.style.filter = filter;
    }

    static addText(image, text) {
        if (image.complete) {
            this.addTextWhenLoaded(image, text);
        } else {
            image.onload = () => {
                this.addTextWhenLoaded(image, text);
            };
        }
    }

    static addTextWhenLoaded(image, text) {
        let parent = image.parentElement;
        let textElement = document.createElement("div");
        textElement.style.fontSize = "0.8em";
        textElement.style.position = "absolute";
        textElement.style.top = "50%";
        textElement.style.left = "50%";
        textElement.style.transform = "translate(-50%, -50%)";
        textElement.classList.add("image-text");
        textElement.innerText = text;
        if(parent)
            parent.appendChild(textElement);
    }

    static async #resize(image, width, height) {
        return new Promise((resolve, reject) => {
            image.onload = () => {
                let hOffset = 0;
                let vOffset = 0;
                if (image.width > image.height) {
                    vOffset = (height - width * image.height / image.width) / 2;
                } else {
                    hOffset = (width - height * image.width / image.height) / 2;
                }

                const canvas = document.createElement('canvas');
                canvas.width = width;
                canvas.height = height;
                const ctx = canvas.getContext('2d');
                ctx.fillStyle = "rgba(0, 0, 0, 0)";
                ctx.fillRect(0, 0, width, height);
                ctx.drawImage(image, hOffset, vOffset, width - hOffset * 2, height - vOffset * 2);
                ctx.canvas.toBlob((blob) => {
                    const resizedImage = new Image();
                    let resizedImageUrl = Thumbnails.createObjectURL(blob);
                    resizedImage.src = resizedImageUrl;
                    resolve(resizedImage);
                }, 'image/png', 0.75);
            }
        });
    }

    static async addCache(key, image) {
        if (!Thumbnails.enableCache)
            return;
        if (Thumbnails.cacheSize > Thumbnails.MAX_CACHE_SIZE)
            Thumbnails.resetCache();
        let blob = await fetch(image.src).then(r => r.blob());
        Thumbnails.cache.set(key,[image,blob.size]);
        Thumbnails.cacheSize += blob.size;
    }

    static async getHash(file) {
        return (await file.getRealPath() + ":" + await file.getLastDateModified());
    }

    static resetCache() {
        let reduceSize = 0;
        let keysToRemove = [];
        for (let key of Thumbnails.cache.keys()) {
            let [image,size] = Thumbnails.cache.get(key);
            if (image != null)
                reduceSize += size;
            if (reduceSize >= Thumbnails.MAX_CACHE_SIZE / 2)
                break;
            keysToRemove.push(key);
        }
        for (let key of keysToRemove) {
            let [image,size] = Thumbnails.cache.get(key);
            Thumbnails.cache.delete(key);
            if (image != null)
                Thumbnails.cacheSize -= size;
            Thumbnails.clearObjectURL(image.src);
        }
        
    }

    static async removeCache(file) {
        let key = await Thumbnails.getHash(file);
        if (Thumbnails.cache.has(key)) {
            if(Thumbnails.cache.get(key) != null) {
                let [image,size] = Thumbnails.cache.get(key);
                Thumbnails.cacheSize -= size;
            }
            Thumbnails.cache.delete(key);
        }
    }

    static async getImageThumbnail(file) {
        let stream = null;
        let image = new Image();
        let blob = null;
        let ms = null;
        try {
            stream = await file.getInputStream();
            ms = new MemoryStream();
            await stream.copyTo(ms);
            blob = new Blob([ms.toArray().buffer]);
            let imageUrl = Thumbnails.createObjectURL(blob);
            image.src = imageUrl;
        } catch (ex) {
            console.error(ex);
        } finally {
            if (ms != null) {
                await ms.close();
            }
            if (stream != null) {
                await stream.close();
            }
        }
        return image;
    }

    static async getHSVFromExtension(extension) {
        let bytes = new TextEncoder().encode(extension);
        let hashValue = await crypto.subtle.digest("SHA-256", bytes);
        let digest = new Uint8Array(hashValue);
        let [r, g, b] = [digest[0] / 256, digest[1] / 256, digest[2] / 256];
        let cmax = Math.max(r, g, b);
        let cmin = Math.min(r, g, b);
        let d = cmax - cmin;
        let h;
        if (cmax == r) h = (60 * ((g - b) / d % 6));
        else if (cmax == g) h = (60 * ((b - r) / d + 2));
        else if (cmax == b) h = (60 * ((r - g) / d + 4));
        let s = (cmax == 0) ? 0 : d / cmax;
        let v = cmax;
        return [h, s, v];
    }

    static createObjectURL(blob) {
        let url = URL.createObjectURL(blob);
        Thumbnails.objectURLs.add(url);
        return url;
    }

    static clearObjectURL(url = null) {
        if(url != null) {
            URL.revokeObjectURL(url);
            Thumbnails.objectURLs.delete(url);
        } else {
            for(let nUrl of Thumbnails.objectURLs) {
                if(nUrl != null)
                    Thumbnails.clearObjectURL(nUrl);
            }
        }
    }
}
